bin/ons_csv_to_ctb_json_read.py

```diff
@@ -73,28 +77,43 @@ def read(self):
 
     def validate_row(self, row, row_num):
         """Validate the fields in a row."""
+        keep_row = True
         for column in self.columns:
             row[column.name] = row[column.name].strip()
 
             if column.required and not row[column.name]:
-                raise ValueError(f'Reading {self.filename}:{row_num} no value supplied '
-                                 f'for required field {column.name}')
+                self.recoverable_error(f'Reading {self.filename}:{row_num} no value supplied '
+                                       f'for required field {column.name}')
+                keep_row = False
+                continue
 
             if column.unique:
                 if row[column.name] in self.unique_column_values[column.name]:
-                    raise ValueError(f'Reading {self.filename}:{row_num} duplicate '
-                                     f'value {row[column.name]} for {column.name}')
+                    self.recoverable_error(f'Reading {self.filename}:{row_num} duplicate '
+                                           f'value {row[column.name]} for {column.name}')
+                    keep_row = False
+                    continue
+
                 self.unique_column_values[column.name].add(row[column.name])
 
             if row[column.name] and column.validate_fn and not \
                     column.validate_fn(row[column.name]):
-                raise ValueError(f'Reading {self.filename}:{row_num} invalid value '
-                                 f'{row[column.name]} for {column.name}')
+                self.recoverable_error(f'Reading {self.filename}:{row_num} invalid value '
+                                       f'{row[column.name]} for {column.name}')
+                if column.required:
+                    keep_row = False
+                    continue
+                row[column.name] = ""
 
-        if self.unique_combo_fields:
+        if self.unique_combo_fields and keep_row:
             combo = tuple([row[f] for f in self.unique_combo_fields])
             if combo in self.unique_combos:
-                raise ValueError(f'Reading {self.filename}:{row_num} duplicate '
-                                 f'value combo {"/".join(combo)} for '
-                                 f'{"/".join(self.unique_combo_fields)}')
-            self.unique_combos.add(combo)
+                self.recoverable_error(f'Reading {self.filename}:{row_num} duplicate '
+                                       f'value combo {"/".join(combo)} for '
+                                       f'{"/".join(self.unique_combo_fields)}')
+
+                keep_row = False
```

Comment:
Again wonder if worth mentioning row will be discarded?

Reply:
This is reported at line 64.